lib/Conversion/BGVToOpenfhe/BGVToOpenfhe.cpp

```diff
@@ -0,0 +1,264 @@
+
+  LogicalResult matchAndRewrite(
+      BinOp op, typename BinOp::Adaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    FailureOr<Value> result = getContextualCryptoContext(op.getOperation());
+    if (failed(result)) return result;
+
+    Value cryptoContext = result.value();
+    rewriter.replaceOp(op,
+                       rewriter.create<OpenfheBinOp>(op.getLoc(), cryptoContext,
+                                                     adaptor.getOperands()[0],
+                                                     adaptor.getOperands()[1]));
+    return success();
+  }
+};
+
+using ConvertNegateOp = ConvertUnaryOp<Negate, openfhe::NegateOp>;
+
+using ConvertAddOp = ConvertBinOp<AddOp, openfhe::AddOp>;
+using ConvertSubOp = ConvertBinOp<SubOp, openfhe::SubOp>;
+using ConvertMulOp = ConvertBinOp<MulOp, openfhe::MulNoRelinOp>;
+
+struct ConvertRotateOp : public OpConversionPattern<Rotate> {
+  ConvertRotateOp(mlir::MLIRContext *context)
+      : OpConversionPattern<Rotate>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      Rotate op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    FailureOr<Value> result = getContextualCryptoContext(op.getOperation());
+    if (failed(result)) return result;
+
+    Value cryptoContext = result.value();
+    auto offsetValue = rewriter.create<arith::ConstantOp>(
+        op.getLoc(), rewriter.getIntegerAttr(rewriter.getIntegerType(64),
+                                             adaptor.getOffset()));
+    rewriter.replaceOp(
+        op, rewriter.create<openfhe::RotOp>(op.getLoc(), cryptoContext,
+                                            adaptor.getX(), offsetValue));
+    return success();
+  }
+};
+
+bool checkRelinToBasis(llvm::ArrayRef<int> toBasis) {
+  if (toBasis.size() != 2) return false;
+  return toBasis[0] == 0 && toBasis[1] == 1;
+}
+struct ConvertRelinOp : public OpConversionPattern<Relinearize> {
+  ConvertRelinOp(mlir::MLIRContext *context)
+      : OpConversionPattern<Relinearize>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      Relinearize op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    FailureOr<Value> result = getContextualCryptoContext(op.getOperation());
+    if (failed(result)) return result;
+
+    auto toBasis = adaptor.getToBasis();
+    if (!checkRelinToBasis(toBasis)) {
+      op.emitError() << "toBasis must be [0, 1], got "
+                     << "[" << toBasis << "]";
+      return failure();
+    }
+
+    Value cryptoContext = result.value();
+    rewriter.replaceOp(op, rewriter.create<openfhe::RelinOp>(
+                               op.getLoc(), cryptoContext, adaptor.getX()));
+    return success();
+  }
+};
+
+bool checkModulusSwitchLevels(unsigned long fromLevel, unsigned long toLevel) {
+  return fromLevel == toLevel + 1;
```

Comment:
Ditto, a comment saying how this is required due to the semantics of the OpenFHE API would be helpful.

Reply:
Applied!